index.js

```diff
@@ -161,178 +88,249 @@ function invalidationsAsPatches(invalidated, currentTree, nextTree) {
   return patches;
 }
 
-Filter.prototype.build = function() {
-  // @ts-ignore
-  let srcDir = this.inputPaths[0];
-  // @ts-ignore
-  let destDir = this.outputPath;
+function invoke(context, fn, args) {
+  return new Promise(resolve => {
+    resolve(fn.apply(context, args));
+  });
+}
+
+class Filter extends Plugin {
+  static shouldPersist(env, persist) {
+    let result;
+
+    if (env.CI) {
+      result = persist && env.FORCE_PERSISTENCE_IN_CI;
+    } else {
+      result = persist;
+    }
 
-  if (this.dependencyInvalidation && !this.dependencies) {
-    this.dependencies = this.processor.initialDependencies(srcDir);
+    return !!result;
   }
 
-  if (this._needsReset) {
+  constructor (inputTree, options) {
+    super([inputTree], {
+      name: (options && options.name),
+      annotation: (options && options.annotation),
+      persistentOutput: true
+    });
+    if (!(this instanceof Filter) || Object.getPrototypeOf(this) === Filter.prototype) {
```

Comment:
I understand the reason we have the the right side of the `||` expression, that provides a nice early failure if someone attempts to `new` this class directly, but what is the point of the left hand side? Is a remnant of pre-class code?